### PR TITLE
Using Middleware

### PR DESCRIPTION
**Pull request by gpltaylor**
Add documentation showing how to implement Middleware when using Kiota.

### CODE REVIEW

**Comment by gpltaylor** [comment]
@microsoft-github-policy-service agree [company="{Fluent Money}"]

**Comment by gpltaylor** [comment]
@microsoft-github-policy-service agree company="Fluent Money"

File: OpenAPI/kiota/middleware.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+date: 21/11/2023
+---
+
+# Implementing Middleware
+By registering custom middleware delegates we can perform operations before a request is made. For example, auditing and filtering the request before we send.
+
+## Middleware
+Create your middleware class and add your business requirements. For example you may wish to add custom Headers to the request or filter headers and content.
+
+```
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:15]:
```suggestion
```cs
```

File: OpenAPI/kiota/middleware.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
+---
+title: Middleware
+description: Learn about implementing middleware into Kiota.
+author: gpltaylor
+ms.topic: overview
+date: 21/11/2023
+---
+
+# Implementing Middleware
+By registering custom middleware delegates we can perform operations before a request is made. For example, auditing and filtering the request before we send.
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:10]:
```suggestion

By registering custom middleware delegates we can perform operations before a request is made. For example, auditing and filtering the request before we send.
```

File: OpenAPI/kiota/middleware.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+## Middleware
+Create your middleware class and add your business requirements. For example you may wish to add custom Headers to the request or filter headers and content.
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:13]:
```suggestion

Create your middleware class and add your business requirements. For example you may wish to add custom Headers to the request or filter headers and content.
```

File: OpenAPI/kiota/middleware.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        string jsonContent = await request.Content.ReadAsStringAsync();
+        Console.WriteLine($"Request: {jsonContent}");
+
+        return await base.SendAsync(request, cancellationToken);
+    }
+}
+```
+## Register Middleware
+Create a Middleware delegate array and use the existing Middleware already implemented within Kiota.HttpClient that includes existing Delegates like Retry, redirect handling and more.
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:29]:
```suggestion

Create a Middleware delegate array and use the existing Middleware already implemented within Kiota.HttpClient that includes existing Delegates like Retry, redirect handling and more.
```

File: OpenAPI/kiota/middleware.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+```
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:31]:
```suggestion
```cs
```

File: OpenAPI/kiota/middleware.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+var delegates = KiotaClientFactory.CreateDefaultHandlers();
+delegates.Add(new SaveRequestHandler());
+```
+Next we create a Delegate chain so middleware is registered in the right order
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:35]:
```suggestion

Next we create a Delegate chain so middleware is registered in the right order.
```

File: OpenAPI/kiota/middleware.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+```
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:37]:
```suggestion
```cs
```

File: OpenAPI/kiota/middleware.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+var handler =
+    KiotaClientFactory.ChainHandlersCollectionAndGetFirstLink(
+        KiotaClientFactory.GetDefaultHttpMessageHandler(),
+        delegates.ToArray());
+```
+
+Finally we create Adapter using our HttpClient that's registered our Middleware. This adapter can then be using when submitting requests. The power of this design means that different adapters/middleware can be used when calling Endpoints and therefore gives flexibility to how and when Middleware is used.
+
+```
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:46]:
```suggestion
```cs
```

File: OpenAPI/kiota/middleware.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+var httpClient = new HttpClient(handler!);
+var adapter = new HttpClientRequestAdapter(authProvider, httpClient:httpClient);
+var client = new PostsClient(adapter);
+```
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:50]:
```suggestion
```

```

File: OpenAPI/kiota/middleware.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
+---
+title: Middleware
+description: Learn about implementing middleware into Kiota.
+author: gpltaylor
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:4]:
```suggestion
author: baywet
ms.author: vibiret
```

This field is used by the bots to assign issues and by our internal ticketing. Not that I don't want you to get credit for your work, but I don't think that set of processes can work with non microsoft employees.

**Review by baywet — CHANGES_REQUESTED** [review]
Thanks for the contribution. Let's start with a couple of linting aspects.

File: OpenAPI/kiota/middleware.md [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        KiotaClientFactory.GetDefaultHttpMessageHandler(),
+        delegates.ToArray());
+```
+
+Finally we create Adapter using our HttpClient that's registered our Middleware. This adapter can then be using when submitting requests. The power of this design means that different adapters/middleware can be used when calling Endpoints and therefore gives flexibility to how and when Middleware is used.
+
+```cs
+var httpClient = new HttpClient(handler!);
+var adapter = new HttpClientRequestAdapter(authProvider, httpClient:httpClient);
+var client = new PostsClient(adapter);
```
Comment (baywet) [inline OpenAPI/kiota/middleware.md:54]:
```suggestion
var client = new PostsClient(adapter);
```

```

**Review by baywet — APPROVED** [review]
thank you so much for the contribution!

**Pull request merged**